### PR TITLE
[WebProfilerBundle] Generate profiler urls with an useful panel

### PR DESCRIPTION
**Pull request by fancyweb**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | no
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

So I did this quickly to get feedback. It could probably be splitted for clarity.

The goal of this work is to provide a better DX with Ajax requests in the WDT.

Having a private profile stack allow us to pass it to the `WebDebugToolbarListener` to process collected data and to directly build the profiler link with the useful panel.

I think that adding a direct link to the dumps panel when there are dumps in an Ajax request would be very useful.

It would also impove the work done in https://github.com/symfony/symfony/pull/26665 by being sure there is actually an exception in the exception panel.

Since all impacted classes were made `@final` in 4.3, it is easy to drop the current behavior (ie remove the double usage of the `$profiles` / `ProfileStack` in the `ProfilerListener` and the optional `ProfileStack` in the `WebDebugToolbarListener`) and to fully switch to the new one in 5.0.

👍 or 👎 ?

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Can you please share a screenshot (with a short animation if that makes sense?) This always helps a lot when improving the profiler.

**Comment by fancyweb** [comment]
Sure, considering there is a `dump` in the code :

Currently : 
screenshot_01

After : 
screenshot_03

If there is a dump, it's very likely that the use wants to see it. It saves one click.

File: src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php [hunk 1]
```diff
@@ -44,15 +47,17 @@ class WebDebugToolbarListener implements EventSubscriberInterface
     protected $mode;
     protected $excludedAjaxPaths;
     private $cspHandler;
+    private $profileStack;
 
-    public function __construct(Environment $twig, bool $interceptRedirects = false, int $mode = self::ENABLED, UrlGeneratorInterface $urlGenerator = null, string $excludedAjaxPaths = '^/bundles|^/_wdt', ContentSecurityPolicyHandler $cspHandler = null)
+    public function __construct(Environment $twig, bool $interceptRedirects = false, int $mode = self::ENABLED, UrlGeneratorInterface $urlGenerator = null, string $excludedAjaxPaths = '^/bundles|^/_wdt', ContentSecurityPolicyHandler $cspHandler = null, ProfileStack $profileStack = null)
```
Comment (fancyweb) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: Since this class is `@final`, can we move this argument to 2nd position as a mandatory one in 5.0 without deprecating anything before? 🤔

**Comment by fancyweb** [comment]
Just added tests.

File: src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: nope, that'd be a hard BC break

**Comment by nicolas-grekas** [comment]
Can we do this using a special debugging header? If yes, wouldn't it make the implementation simpler?

**Comment by fancyweb** [comment]
We can technically do it with special headers. It heavily reduces the amount of added code. 
However I see at least three problems : 

1. Adding a random header in the collectors looks bad. It adds this logic in the `HttpKernel` component, thus polluting the component while it's just for us and our web profiler bundle.
2. By using a header, we lose the control of which panels can be accessed directly : any 3rd party collector can force the redirection to its panel if it overrides the header and is "collected" last.
3. Managing the priority of the preferred panel (exception > all) becomes hard because collectors can be processed in any order + their name is dynamic (the collector name === the panel name)

**Comment by fancyweb** [comment]
Friendly ping @nicolas-grekas, @ro0NL, @BoShurik, @ismail1432 and @fabpot (since you resolved a discussion), is this going anywhere? I deeply think it's useful and would like to see this in 4.4.

The only thing left to do is to manage the constructor arguments of `WebDebugToolbarListener` to deprecate not passing a `ProfileStack` as a 2nd mandatory argument, although I don't understand why it's needed since the class is `@final` since Symfony 4.3 and according to https://symfony.com/doc/current/contributing/code/bc.html that means it is really final in the next major version (and so the promise does not apply anymore).

File: src/Symfony/Component/HttpKernel/EventListener/ProfilerListener.php [hunk 2]
```diff
@@ -94,9 +97,13 @@ public function onKernelResponse(FilterResponseEvent $event)
         $this->profiles[$request] = $profile;
 
         $this->parents[$request] = $this->requestStack->getParentRequest();
+
+        if ($this->profileStack instanceof ProfileStack) {
```
Comment (stof) [inline src/Symfony/Component/HttpKernel/EventListener/ProfilerListener.php]: should this be done also for subrequests, or only for master requests ?

File: src/Symfony/Component/HttpKernel/Profiler/ProfileStack.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    public function has(Request $request): bool
+    {
+        return isset($this->profiles[$request]);
+    }
+
+    public function get(Request $request): Profile
+    {
+        return $this->profiles[$request];
```
Comment (stof) [inline src/Symfony/Component/HttpKernel/Profiler/ProfileStack.php]: the implementation does not deal with cases where the request does not have a profile, which would break with a TypeError (which would then be considered as a bug in Symfony by users, and rightfully so). You should check this case and throw a meaningful exception.

File: src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php [hunk 4]
```diff
@@ -65,11 +70,33 @@ public function onKernelResponse(FilterResponseEvent $event)
...
+            } else {
+                $token = $response->headers->get('X-Debug-Token');
+            }
             try {
                 $response->headers->set(
                     'X-Debug-Token-Link',
-                    $this->urlGenerator->generate('_profiler', ['token' => $response->headers->get('X-Debug-Token')], UrlGeneratorInterface::ABSOLUTE_URL)
+                    $this->urlGenerator->generate('_profiler', [
+                        'token' => $token,
+                        'panel' => $panel ?? null,
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: you should define `$panel` as `null` at the beginning of the logic instead of dealing with an undefined variable here.

File: src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php [hunk 5]
```diff
@@ -65,11 +70,33 @@ public function onKernelResponse(FilterResponseEvent $event)
...
+                $profile = $this->profileStack->get($request);
+
+                $token = $profile->getToken();
+
+                foreach ($profile->getCollectors() as $collector) {
+                    if ($collector instanceof ExceptionDataCollector && $collector->hasException()) {
+                        $panel = $collector->getName();
+
+                        break;
+                    } elseif ($collector instanceof DumpDataCollector && $collector->getDumpsCount() > 0) {
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: `if` is enough, as the previous `if` stops execution of this branch of the code with `break` already.

**Comment by stof** [comment]
I also suggest rewording the PR title (which will appear in generated changelogs) and the commit message to talk about the feature being implemented. The profile stack is not the feature, but an internal implementation detail (proof is that the class is `@internal`)

File: src/Symfony/Component/HttpKernel/EventListener/ProfilerListener.php (hunk 2 above)
Comment (fancyweb) [reply src/Symfony/Component/HttpKernel/EventListener/ProfilerListener.php]: Thanks for your review. The exact same profiles that are currently added to the `$profiles` class property are added to the stack. So the behavior is not modified. So it add them depending of the "only master requests" arg.

File: src/Symfony/Component/HttpKernel/Profiler/ProfileStack.php (hunk 3 above)
Comment (fancyweb) [reply src/Symfony/Component/HttpKernel/Profiler/ProfileStack.php]: The class is `@internal`. It's our responsibility to use `get()` after `has()` (that's what is done). Isn't it enough? 🤔

Comment (stof) [reply src/Symfony/Component/HttpKernel/Profiler/ProfileStack.php]: I would still make the code report a meaningful error if you misuse it rather than a TypeError. Especially given that it does not either document the responsibility it expects from us (and so future contributors may not know it before using it).
I'm fine with not adding doc explaining that if the failure mode is clear. But a TypeError saying we cannot return `null` or a failure saying that the item does not exist in the SplObjectStorage (I don't remember how SplObjectStorage deals with that) is not a clear failure mode.

**Comment by javiereguiluz** [comment]
@fancyweb thanks for working on this! About the new feature, I like it because it's a nice little DX improvement.

About the technical details, I don't have a strong opinion. I'd like to add less code to implement this ... but [this comment](https://github.com/symfony/symfony/pull/32491#issuecomment-517301045) shows some compelling arguments against that ... so I'm divided.

**Comment by nicolas-grekas** [comment]
I still feel like that's a lot of code for the use case.
Alternative idea: link to `?panel=auto` and let that trigger a redirect to the best panel for the corresponding profile. Up to us to implement what "best" means here, eg the dump, or the error page, etc.

**Comment by fancyweb** [comment]
@nicolas-grekas Your alternative idea seems very good actually. I will give it a try quickly.

**Comment by stof** [comment]
I suggest closing this PR in favor of https://github.com/symfony/symfony/pull/33783, which has several advantages IMO:

- it adds logic when someone clicks on the link, rather than adding more work in the listener running for all responses
- the listener runs before the profile is complete (many collectors are collecting things only during `kernel.terminate` thanks to `lateCollect`), meaning that its guessing is based on incomplete data. On the other hand, the controller displaying a panel has the full data of the profile.

**Comment by stof** [comment]
and btw, putting it in the controller also allows using this guessing in other links (we could make search results use it too).
